More Mount Conflict Detection

otakup0pe opened the pull request:
Ensure that we are not creating a "submount" underneath an already
existing mountpoint. This (hopefully) fixes #2878

Note that (locally) I'm seeing intermittent (~25% of the time on the full test suite) build failures 😿  

```
--- FAIL: TestCluster_ForwardRequests (9.16s)
        cluster_test.go:350: err: error during forwarding RPC request
```

otakup0pe commented:
I'm missing something about the intersection between the router and logical backends... Doing some (manual) smoke testing, I noticed this repeating in the logs...

```
2017/06/24 10:14:45.630510 [ERROR] rollback: error rolling back: path=foo/ error=unsupported path
```

jefferai commented:
The issue is that what's in the router's `Mount` code is only an extra check; the code in mount.go and auth.go check for an exact match via the router's MatchingMount, but then if that doesn't find a conflict they add the new entry into the table, *then* mount it on the router. The reason is that once a backend is able to be sent requests, we want it to already be stored so if Vault goes down it can come back up.

The right thing to do would probably be to take the logic you wrote and the check above for a longest prefix and combine it into two functions inside the router, something like AllowedMount and allowedMountInternal. The latter takes that logic and returns an error or not, and can be used inside the router's Mount function. The former does the same thing but gets a read lock (see MatchingMount); this can be called from mount.go and auth.go instead of the existing MatchingMount call.

otakup0pe commented:
Thanks for the guidance, those changes cleared up the rollback error. Any guidance on why postgres tests are (consistently) failing? Basis test suite (`make test TEST=./vault`) is fine.

```
--- FAIL: TestPostgreSQL_RenewUser (8.30s)
        postgresql_test.go:182: Could not connect with new credentials: pq: password authentication failed for user "v-test-test-ZLS5aFHTd2RNsz8eDKV8-1498407266"
--- FAIL: TestPostgreSQL_RevokeUser (7.56s)
        postgresql_test.go:250: Could not connect with new credentials: pq: password authentication failed for user "v-test-test-WzseS2eTjqgckXSBWefZ-1498407274"
FAIL
FAIL    github.com/hashicorp/vault/plugins/database/postgresql  36.679s
```

jefferai commented on vault/auth.go:
```diff
@@ -73,7 +73,7 @@ func (c *Core) enableCredential(entry *MountEntry) error {
 	}
 
-	if match := c.router.MatchingMount(credentialRoutePrefix + entry.Path); match != "" {
+	if match := c.router.MountConflict(credentialRoutePrefix + entry.Path); match != "" {
```
For ease of understanding, can you change `match` here to `conflict` and in the error message say "conflicting mount at ..."?

jefferai commented on vault/router.go:
```diff
@@ -60,6 +60,13 @@ func (r *Router) Mount(backend logical.Backend, prefix string, mountEntry *Mount
+	// with an existing mountpoint
+	if prefix != "" {
+		if match := r.MatchingPrefix(prefix); match != "" {
+			return fmt.Errorf("cannot mount under existing mount '%s'", match)
```
Isn't this checking for mounting *over* an existing mount?

jefferai commented on vault/router.go:
```diff
@@ -174,6 +181,32 @@ func (r *Router) MatchingMount(path string) string {
+func (r *Router) MatchingPrefix(path string) string {
+	var existing string = ""
+	fn := func(existing_path string, _v interface{}) bool {
+		fmt.Println(fmt.Sprintf("comparing %v %v", path, existing_path))
```
I'm guessing this was a debug line you meant to take out

otakup0pe replied:
How embarrassing 😬

jefferai replied:
Mostly it was like, um, that's going to spam logs a *LOT*  :-)

otakup0pe replied:
Which is why it's embarrassing 👼 

New commit incorporates the feedback and I added a little blurb to the docs about mount conflicts.

jefferai commented on vault/router.go:
```diff
@@ -174,6 +181,31 @@ func (r *Router) MatchingMount(path string) string {
 }
 
+//MatchingPrefix returns a mount prefix that a path may be a part of
+func (r *Router) MatchingPrefix(path string) string {
```
This function doesn't get a read lock but should, as it can be called externally (via MountConflict from the mount code). However, it is called with a lock already held from the router's Mount function.

The way to deal with this would be to have MatchingPrefix grab a read lock, then call matchingPrefixInternal which has the actual logic. The router's Mount function would then just call matchingPrefixInternal directly.

jefferai commented on website/source/docs/secrets/index.html.md:
```diff
@@ -59,6 +59,13 @@ Once a secret backend is mounted, you can interact with it directly
 at its mount point according to its own API. You can use the `vault path-help`
 system to determine the paths it responds to.
 
+Note that mount point cannot conflict with each other in Vault. There are
```
s/point/points

jefferai commented on vault/router.go:
```diff
@@ -174,6 +181,31 @@ func (r *Router) MatchingMount(path string) string {
+//MatchingPrefix returns a mount prefix that a path may be a part of
+func (r *Router) MatchingPrefix(path string) string {
+	var existing string = ""
+	fn := func(existing_path string, _v interface{}) bool {
```
If this function executes _at all_ wouldn't it mean that we've found a prefix? Is the `if` check unnecessary here?

otakup0pe commented:
Added the read lock and fixed a typo but wasn't clear on the third statement. The `if` in what is now `matchingPrefixInternal` checks each entry in the radix tree... If you meant the `if` on line [67](https://github.com/otakup0pe/vault/blob/f3136f0fa49d5c32e5f63c103bb1bb0a72210e6e/vault/router.go#L67) - that is there because some entities (like policies iirc) didn't seem to have a prefix set...

otakup0pe commented:
Just noticed tests failed (they are passing locally). Will take a look this afternoon/evening... Unless someone can restart the build 👼

jefferai commented on vault/router.go:
```diff
@@ -62,6 +62,13 @@ func (r *Router) Mount(backend logical.Backend, prefix string, mountEntry *Mount
 	}
+	// If this is a secret backend, check to see if the prefix conflicts
+	// with an existing mountpoint
+	if prefix != "" {
```
I'm pretty sure that we don't allow empty prefixes in the API calls or core mount function. Have you seen this? This `if` seems unnecessary.

chrishoffman replied:
Yes, this check appears unnecessary.

jefferai commented on vault/router.go:
```diff
@@ -205,6 +212,39 @@ func (r *Router) MatchingMount(path string) string {
+
+// MountConflict determines if there are potential path conflicts
+func (r *Router) MountConflict(path string) string {
+	if exact_match := r.MatchingMount(path); exact_match != "" {
```
Really this function should grab a single read lock for the entire function and then call matchingMountInternal and matchingPrefixInternal. That way nothing can change between the two functions.

jefferai commented:
Regarding the `if` -- both, actually. I left a comment on the one at line 67. The other one was about the use of HasPrefix in the walking function. Wouldn't we never even hit that part of the walk function if HasPrefix wasn't true?

otakup0pe commented:
Shuffled around the "internal" functions. When I removed the prefix matching code at in `Mount`, the build fails and I see a bunch of warnings/errors about audit logs (not sure why I thought policies). They might be a red herring, there is a lot of output to parse when the vault test suite fails.

```
2017/07/26 22:54:14.121014 [ERROR] audit: backend failed to log request: backend=foo error=failed
2017/07/26 22:54:14.121028 [ERROR] audit: backend failed to log request: backend=foo error=failed
2017/07/26 22:54:14.121037 [ERROR] audit: backend failed to log request: backend=bar error=failed
2017/07/26 22:54:14.121166 [ERROR] audit: backend failed to log response: backend=foo error=failed
2017/07/26 22:54:14.121178 [ERROR] audit: backend failed to log response: backend=foo error=failed
2017/07/26 22:54:14.121185 [ERROR] audit: backend failed to log response: backend=bar error=failed
2017/07/26 22:54:14.121376 [ERROR] audit: backend failed to log request: backend=foo error=failed
2017/07/26 22:54:14.121403 [ERROR] audit: backend failed to log request: backend=foo error=failed
2017/07/26 22:54:14.121430 [ERROR] audit: backend failed to log request: backend=bar error=failed
```

```
2017/07/26 22:54:14.078248 [ERROR] core: failed to create audit entry: path=noop/ error=failing enabling
2017/07/26 22:54:14.078259 [ERROR] core: failed to create audit entry: path=noop/ error=failing enabling
2017/07/26 22:54:14.078268 [ERROR] core: failed to create audit entry: path=noop2/ error=failing enabling
```

jefferai commented:
@otakup0pe Your build works fine -- can you explain what you mean about "when I removed the prefix matching code in Mount"?

jefferai commented:
Rather, the CI tests pass fine...so not sure what errors you're meaning.

otakup0pe commented:
If I remove lines 67-71 the build will repeatedly fail locally. I left them in for the PR because it seems to be working properly based on the tests and the manual testing I did... I can push without those lines and we can see what Travis says if you'd like.

chrishoffman commented on vault/router.go:
```diff
@@ -62,6 +62,13 @@ func (r *Router) Mount(backend logical.Backend, prefix string, mountEntry *Mount
+	// If this is a secret backend, check to see if the prefix conflicts
+	// with an existing mountpoint
+	if prefix != "" {
+		if match := r.matchingPrefixInternal(prefix); match != "" {
```
I'm a little concerned about adding this check in the router.  I've confirmed that adding this check will prevent servers that are in the situation of overlapping mounts from unsealing, where they would start fine before.  This may be ok since vault will never unseal and you can just downgrade to the earlier version, move their mounts, and then proceed with the upgrade.  It's at the very least something to not in the upgrade docs.  @jefferai thoughts?

otakup0pe replied:
Let me know how to proceed. I can add some context on the upgrade docs, or is handling the data migration something we want to consider?

jefferai replied:
I think my preference here would be to allow it in the router but not the mount code. So existing setups will work, but it won't be allowed for new mounts.

chrishoffman commented on vault/router.go:
```diff
@@ -197,14 +204,48 @@ func (r *Router) MatchingMountByAccessor(mountAccessor string) *MountEntry {
+func (r *Router) MountConflict(path string) string {
+	r.l.RLock()
+	if exact_match := r.matchingMountInternal(path); exact_match != "" {
+		r.l.RUnlock()
```
You should just `defer` the `RUnlock` instead of unlocking on all return paths.

jefferai commented:
Note: moved to next release as it's super late in the dev cycle for 0.8 and this is a very core change, so that will give more time for review and baking the change in tests and such.

jefferai commented:
@otakup0pe Just wanted to check on on this and see if you were going to address the remaining review items.

otakup0pe commented:
Hi @jefferai I was off grid for several weeks. Will pick wrap this up now that I'm back!

otakup0pe commented:
I think this addresses the feedback. I adjusted the tests to account for the new pattern.

vishalnayak commented on vault/router_test.go:
```diff
@@ -162,6 +167,24 @@ func TestRouter_Mount(t *testing.T) {
+		t.Fatalf("bad: prod/aws/")
+	}
+
+	err = r.Mount(n, "prod/", subMountEntry, view)
```
Can we have a comment here as to why this should pass?

vishalnayak commented on vault/router_test.go:
```diff
@@ -118,6 +118,11 @@ func TestRouter_Mount(t *testing.T) {
 		t.Fatalf("err: %v", err)
 	}
 
+	meUUID, err = uuid.GenerateUUID()
```
Can we move this to be placed just above the definition of subMountEntry?

vishalnayak commented on vault/router.go:
```diff
@@ -201,14 +200,46 @@ func (r *Router) MatchingMountByAccessor(mountAccessor string) *MountEntry {
 func (r *Router) MatchingMount(path string) string {
 	r.l.RLock()
-	mount, _, ok := r.root.LongestPrefix(path)
+	var mount = r.matchingMountInternal(path)
```
[Nitpick][Optional] We could do a `defer r.l.RUnlock()` and  `return r.matchingMountInternal(path)` and avoid the local var `mount`.
